File: archive/v/visual-basic/roman-numeral.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq

Module RomanNumeral

    Private Function GetValue(c As Char) As Integer
        Select Case c
            Case "M"c : Return 1000
            Case "D"c : Return 500
            Case "C"c : Return 100
            Case "L"c : Return 50
            Case "X"c : Return 10
            Case "V"c : Return 5
            Case "I"c : Return 1
            Case Else
                Throw New ArgumentException("Invalid Roman numeral character")
        End Select
    End Function

    Private Function RomanToDecimal(roman As String) As Integer
        Dim total As Integer = 0
        Dim lastValue As Integer = 0

        For i As Integer = roman.Length - 1 To 0 Step -1
            Dim currentValue As Integer = GetValue(roman(i))

            If currentValue < lastValue Then
                total -= currentValue
            Else
                total += currentValue
            End If

            lastValue = currentValue
        Next

        Return total
    End Function

    Sub Main(args As String())
        If args.Length = 0 OrElse String.IsNullOrWhiteSpace(args(0)) Then
            Console.WriteLine("Usage: please provide a string of roman numerals")
            Return
        End If

        Try
            Console.WriteLine(RomanToDecimal(args(0).ToUpperInvariant()))
        Catch ex As ArgumentException
            Console.WriteLine("Error: invalid string of roman numerals")
        End Try
    End Sub

End Module

File: archive/p/perl/roman-numeral.pl
#!/usr/bin/env perl
use v5.42;

my %VAL = (
    I => 1,
    V => 5,
    X => 10,
    L => 50,
    C => 100,
    D => 500,
    M => 1000,
);

sub usage   { say "Usage: please provide a string of roman numerals"; exit }
sub invalid { say "Error: invalid string of roman numerals";          exit }

my ($s) = @ARGV;

defined $s or usage();

$s =~ s/\A\s+|\s+\z//g;
$s eq '' and say 0 and exit;

$s =~ /\A[IVXLCDM]+\z/ or invalid();

my ( $total, $prev ) = ( 0, 0 );

for my $c ( reverse split //, $s ) {
    my $v = $VAL{$c};
    $total += $v < $prev ? -$v : $v;
    $prev = $v;
}

say $total;
